lib/oli/delivery/paywall/providers/stripe.ex

```diff
@@ -68,40 +71,43 @@ defmodule Oli.Delivery.Paywall.Providers.Stripe do
       )
       |> URI.encode_query()
 
-    private_secret = Application.fetch_env!(:oli, :stripe_provider)[:private_secret]
-
-    headers = [
-      Authorization: "Bearer #{private_secret}",
-      "Content-Type": "application/x-www-form-urlencoded"
-    ]
-
-    case http().post(
-           "https://api.stripe.com/v1/payment_intents",
-           body,
-           headers
-         ) do
-      {:ok, %{status_code: 200, body: body}} ->
-        intent = Poison.decode!(body)
-
-        %{"id" => id} = intent
-
-        attrs = %{
-          type: :direct,
-          generation_date: DateTime.utc_now(),
-          amount: amount,
-          provider_payload: intent,
-          provider_id: id,
-          provider_type: :stripe,
-          section_id: product.id
-        }
-
-        case Oli.Delivery.Paywall.create_pending_payment(user, section, attrs) do
-          {:ok, _} -> {:ok, intent}
-          e -> e
-        end
+    headers = fn private_secret ->
+      [
+        Authorization: "Bearer #{private_secret}",
+        "Content-Type": "application/x-www-form-urlencoded"
+      ]
+    end
 
-      _ ->
+    attrs = fn intent ->
+      %{
+        type: :direct,
+        generation_date: DateTime.utc_now(),
+        amount: section.amount,
+        provider_payload: intent,
+        provider_id: intent["id"],
+        provider_type: :stripe,
+        section_id: section.id
+      }
+    end
+
+    with {:ok, [_, {:private_secret, private_secret}]} <-
+           Application.fetch_env(:oli, :stripe_provider),
+         {:ok, %{status_code: 200, body: body}} <-
+           http().post(
+             "https://api.stripe.com/v1/payment_intents",
+             body,
+             headers.(private_secret)
+           ),
+         {:ok, intent} <- Poison.decode(body),
+         {:ok, _} <- Oli.Delivery.Paywall.create_pending_payment(user, section, attrs.(intent)) do
+      {:ok, intent}
+    else
+      {:ok, %HTTPoison.Response{} = response} ->
+        Logger.error("Failed request to Stripe payment intents, response: #{inspect(response)}")
```

Comment:
Re-organized this to be defensive when using `Application.fetch_env`, if the previous `Application.fetch_env!` did raise an exception, it wouldn't be handled properly.